flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/sink/ConstraintValidator.java

```diff
@@ -0,0 +1,201 @@
+        if (processedElement != null) {
+            output.collect(processCharConstraint(processedElement));
+        }
+    }
+
+    private StreamRecord<RowData> processNotNullConstraint(StreamRecord<RowData> element) {
+        if (notNullEnforcer == null) {
+            return element;
+        }
+
+        final RowData rowData = element.getValue();
+        for (int index : notNullFieldIndices) {
+            if (rowData.isNullAt(index)) {
+                switch (notNullEnforcer) {
+                    case ERROR:
+                        throw new TableException(
+                                String.format(
+                                        "Column '%s' is NOT NULL, however, a null value is being written into it. "
+                                                + "You can set job configuration '%s'='drop' "
+                                                + "to suppress this exception and drop such records silently.",
+                                        allFieldNames[index],
+                                        TABLE_EXEC_SINK_NOT_NULL_ENFORCER.key()));
+                    case DROP:
+                        return null;
+                }
+            }
+        }
+        return element;
+    }
+
+    private StreamRecord<RowData> processCharConstraint(StreamRecord<RowData> element) {
+        if (charPrecisionEnforcer == null) {
+            return element;
+        }
+
+        final RowData rowData = element.getValue();
+        final List<Tuple2<Integer, String>> newStrings = new ArrayList<>();
+        for (Tuple2<Integer, Integer> tuple : charFields) {
+            final int fieldIdx = tuple.f0;
+            final int length = tuple.f1;
+            final String stringValue = rowData.getString(fieldIdx).toString();
+
+            if (stringValue.length() > length) {
+                switch (charPrecisionEnforcer) {
+                    case ERROR:
+                        throw new TableException(
+                                String.format(
+                                        "Column '%s' is of CHAR/VARCHAR type with a length of '%s', however, "
+                                                + "a string value with a length of '%s' is being written into it. "
+                                                + "You can set job configuration '%s'='trim' to suppress this exception "
+                                                + "and trim such string values to the length defined by the CHAR/VARCHAR "
+                                                + "type.",
+                                        allFieldNames[fieldIdx],
+                                        length,
+                                        stringValue.length(),
+                                        TABLE_EXEC_SINK_CHAR_PRECISION_ENFORCER.key()));
+                    case TRIM:
+                        newStrings.add(new Tuple2<>(fieldIdx, stringValue.substring(0, length)));
```

Comment:
`new Tuple2<>` is expensive, can't we perform the trimming in place already? We could use a reusable `RowData` instance for buffering and copy over the values in `UpdatableRowData` later.

Reply:
We can do it here and not use a `List<Tuple2<>>`, but how can we reuse the `RowData`? that's the `RowData` iface so we don't know the implementation and call directly `.setField()` on it. Could you please provide more details, probably I'm missing something.